### PR TITLE
Adds MicroProfile Metrics to Hikari CP

### PR DESCRIPTION
**Pull request by ljnelson**
Signed-off-by: Laird Nelson <laird.nelson@oracle.com>

### CODE REVIEW

**Comment by ljnelson** [comment]
@tjquinno A first sketch. See what you think.

**Comment by tjquinno** [comment]
A few thoughts.

Looks very good and nicely tracks the Micrometer one you linked to. 

1. Re: the `MicroProfileMetricsTracker#close` method.
    1.  IIUC the class is instantiated once per pool. The metrics it creates are tagged with the pool name. It looks as if the `close` method use the name _and the tag_ in selecting metrics for removal.
    2. It turns out the `removeMatching` implementation iterates over all metrics in the registry, whereas `remove(MetricID)` is much more efficient. If pool closing is rare, then `removeMatching` is probably fine. If this happens frequently, maybe collect the `MetricID`s of the metrics created for the pool and then `close` can remove each one individually.

2. The Micrometer implementation differs from the other document by Brett. I mention this because this issue specifically linked to that document and so might reflect some expectations about how our implementation might look. That said, the running Micrometer code is probably even more forceful in setting expectations!
    1.  The document suggested using a `Histogram` for pool usage. The Micrometer implementation did not do that. Might be worth considering if the histogram's data would be useful.
    2. The document builds the pool name into the metric name. The Micrometer implementation uses tags instead. The tagging approach seems better to me.

**Comment by ljnelson** [comment]
Thanks @tjquinno; good suggestions.  I'm not sure whether to use the histogram or not given that the Micrometer support doesn't use it.  I'll leave things as-is for the moment, and apply your other suggestions (particularly the one about including the tags).  If you know something about histogram support that would make it a slam dunk for sure, I'll reconsider.

**Comment by tjquinno** [comment]
In my mind, it's a usefulness question on the histogram vs. simple timer choice for the connection usage, but certainly not a slam dunk.

The histogram would let the user see the distribution of how long connections in the pool were in use (as well as min, max, mean, total, std dev of those usage times), vs. the simple timer which would report a total count of how many uses of a pooled connection occurred and an accumulated time over all of those connection usages of how long they were in use. 

TBH, I'm not sure as a user exactly how I'd use either the histogram or the simple timer output to influence how I wanted to configure the pool, but the "thing" that the histogram measures and reports seems much more understandable and potentially more useful to me than the odd accumulation of connection usages offered by the simple timer.

**Comment by ljnelson** [comment]
@tjquinno Let's see if I'm still on the right track here.  I followed the Wiki document which is clearly out of date (it lists a small subset of supported metrics) but I think the spirit, as you say, is right on.  I'll need to circle back and make sure my names line up with it (I don't think the Micrometer ones, which I used as a starting point, do).

I incorporated the remove-by-key semantics in the `close()` method.  Anything else ginormous jump out at you?

**Comment by ljnelson** [comment]
Torn on the names; on the one hand, we have names [like this](https://github.com/brettwooldridge/HikariCP/blob/f9b2c3d372b3f28911a872191ad62953c84bafa4/src/main/java/com/zaxxer/hikari/metrics/dropwizard/CodaHaleMetricsTracker.java#L38-L48) that clearly were drawn from the wiki.

And in the Prometheus area we have names [like this](https://github.com/brettwooldridge/HikariCP/blob/f9b2c3d372b3f28911a872191ad62953c84bafa4/src/main/java/com/zaxxer/hikari/metrics/prometheus/PrometheusMetricsTracker.java#L33-L46).

And in the Micrometer area we have names [like this](https://github.com/brettwooldridge/HikariCP/blob/f9b2c3d372b3f28911a872191ad62953c84bafa4/src/main/java/com/zaxxer/hikari/metrics/micrometer/MicrometerMetricsTracker.java#L36-L49).

If I were to follow the spirit of the wiki page, I'd use the DropWizard flavor.  Otherwise I think I'd use the Micrometer flavor (for no particularly good reason).  What do you think?

**Comment by tjquinno** [comment]
An aside: In reading the page at the Prometheus link, I see that they use a `Summary` metric (which is close to the MP `Histogram`) for several metrics, not just usage. Using `Histogram` in our implementation for all of those `Summary` metrics would deviate further from the existing Micrometer implementation you linked to, but maybe that's OK if not better. Judgement call about what users might find most useful. 

As for the names, what looks inconsistent between the Micrometer ones and the Prometheus ones are (I think, at a glance) consistent. Prometheus format calls for replacing dots with underscores in the names. The Helidon impl for Prometheus output (in compliance with Prometheus rules) automatically appends the units for the metric (if one other than `NONE` is set). So I think the names you've got seem fine.

**Comment by ljnelson** [comment]
OK; any thoughts on the `connections` prefix used by the Micrometer one and absent everywhere else?  And the capitalization or lack of it?

**Comment by tjquinno** [comment]
Not sure I understand. The Prometheus implementation uses `hikaricp_connection_xxx`. The Micrometer one uses `hikaricp.connections.xxx`(plural "connections" instead of singular but otherwise consistent given the formatting conventions). 

I would avoid capitalizing the metric names as Brett's write-up illustrates. Lower-case seems to be _de riguer_ in what I've seen, anyway. Certainly the Helidon built-in vendor and base metrics names are lower-case.

**Comment by ljnelson** [comment]
OK, so lower case it is.  I cited the Prometheus names but meant the DropWizard ones (what, can't you read my mind?).  For example, DropWizard has [`Wait`](https://github.com/brettwooldridge/HikariCP/blob/f9b2c3d372b3f28911a872191ad62953c84bafa4/src/main/java/com/zaxxer/hikari/metrics/dropwizard/CodaHaleMetricsTracker.java#L39).  When [smushed together](https://github.com/brettwooldridge/HikariCP/blob/f9b2c3d372b3f28911a872191ad62953c84bafa4/src/main/java/com/zaxxer/hikari/metrics/dropwizard/CodaHaleMetricsTracker.java#L54) with the pool name and the category ([`pool`](https://github.com/brettwooldridge/HikariCP/blob/f9b2c3d372b3f28911a872191ad62953c84bafa4/src/main/java/com/zaxxer/hikari/metrics/dropwizard/CodaHaleMetricsTracker.java#L38)), you presumably get something like `frob.pool.Wait`.  Finally, there is the bizarre thing where they're [measuring threads but calling them connections](https://github.com/brettwooldridge/HikariCP/blob/f9b2c3d372b3f28911a872191ad62953c84bafa4/src/main/java/com/zaxxer/hikari/metrics/dropwizard/CodaHaleMetricsTracker.java#L68-L69), and that seems to be true throughout the various implementations.

So: I think we're saying: follow the Micrometer naming convention, keep things lowercase, and perpetuate the thread/connection confusion because maybe there's some reason that exists.  Does that sound right?

**Comment by tjquinno** [comment]
SGTM! 

I can talk myself into believing that the thread/connection conflation maybe makes a little sense. From HikariCP's connection-centric point of view, there are "things" that want connections but cannot have them (yet). The outward metric for that, therefore, shows "pending connections." 

The fact that -- today -- threads are the direct measurable "things" that can be waiting to get a connection causes HIkariCP to measure those internally but display "pending connections." Maybe someday there might be other kinds of "things" waiting for connections instead/as well, but Hikari assumes the human consumer of the metrics won't care whether it's threads or fibers or strands or filaments -- just that the demand for connections exceeds the supply.

It just so happens that, today, measuring the waiting threads is _the_ way to measure pending connections.

Or am I giving them too much credit!?

**Comment by ljnelson** [comment]
@tjquinno One more round then I'll un-draft it.

**Comment by tjquinno** [comment]
Go for undrafting.

File: integrations/cdi/datasource-hikaricp/src/main/java/io/helidon/integrations/datasource/hikaricp/cdi/MicroProfileMetricsTracker.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+    private static final String METRIC_NAME_WAIT = HIKARI_METRIC_NAME_PREFIX + "connections.wait";
+
+    private static final String METRIC_NAME_USAGE = HIKARI_METRIC_NAME_PREFIX + "connections.usage";
+
+    private static final String METRIC_NAME_CONNECT = HIKARI_METRIC_NAME_PREFIX + "connections.creation";
+
+    private static final String METRIC_NAME_TIMEOUT_RATE = HIKARI_METRIC_NAME_PREFIX + "connections.timeout";
+
+    private static final String METRIC_NAME_TOTAL_CONNECTIONS = HIKARI_METRIC_NAME_PREFIX + "connections";
```
[the remark this answers is not in the document]
Comment (ljnelson) [reply integrations/cdi/datasource-hikaricp/src/main/java/io/helidon/integrations/datasource/hikaricp/cdi/MicroProfileMetricsTracker.java:45]: I was following https://github.com/brettwooldridge/HikariCP/blob/ed2da5f1f4ef19f871fac12effc0b199706905dc/src/main/java/com/zaxxer/hikari/metrics/micrometer/MicrometerMetricsTracker.java#L45. WDYT?

**Comment by ljnelson** [comment]
Note in this latest (squashable) commit I also made the MicroProfile Metrics module requirement `static`, since it does not have to be present at runtime for functionality to work.

File: integrations/cdi/datasource-hikaricp/src/main/java/io/helidon/integrations/datasource/hikaricp/cdi/MicroProfileMetricsTracker.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (tjquinno) [reply integrations/cdi/datasource-hikaricp/src/main/java/io/helidon/integrations/datasource/hikaricp/cdi/MicroProfileMetricsTracker.java:45]: Yeah, that's probably fine.

**Review by tjquinno — APPROVED** [review]
Make it so.

**Pull request merged**